File: src/ruby/tormenta20/models/poder.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing a Poder (Power/Ability) in Tormenta20.
    #
    # Powers can be unique origin abilities, deity-granted powers, Tormenta powers, etc.
    #
    # @example Find deity-granted powers
    #   Poder.poderes_concedidos
    #
    # @example Find powers by type
    #   Poder.by_type("poder_tormenta")
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier
    # @!attribute [rw] name
    #   @return [String] Power name
    # @!attribute [rw] type
    #   @return [String] Power type (see TYPES constant)
    # @!attribute [rw] description
    #   @return [String] Power description
    # @!attribute [rw] effects
    #   @return [Array] Power effects
    # @!attribute [rw] prerequisites
    #   @return [Array] Prerequisites to acquire this power
    # @!attribute [rw] origin_id
    #   @return [String, nil] Associated origin ID (for unique origin powers)
    # @!attribute [rw] deities
    #   @return [Array<String>] Deities that grant this power
    class Poder < Base
      self.table_name = "poderes"
      self.inheritance_column = nil

      # Valid power types.
      #
      # @return [Array<String>] List of valid type values
      TYPES = %w[
        habilidade_unica_origem
        poder_concedido
        poder_tormenta
        poder_classe
        poder_geral
        poder_combate
        poder_destino
        poder_magia
      ].freeze

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true
      validates :type, presence: true, inclusion: { in: TYPES }

      # @!group Scopes

      # @!method habilidades_unicas
      #   Filter unique origin abilities.
      #   @return [ActiveRecord::Relation<Poder>]
      scope :habilidades_unicas, -> { where(type: "habilidade_unica_origem") }

      # @!method poderes_concedidos
      #   Filter deity-granted powers.
      #   @return [ActiveRecord::Relation<Poder>]
      scope :poderes_concedidos, -> { where(type: "poder_concedido") }

      # @!method poderes_tormenta
      #   Filter Tormenta powers.
      #   @return [ActiveRecord::Relation<Poder>]
      scope :poderes_tormenta, -> { where(type: "poder_tormenta") }

      # @!method poderes_classe
      #   Filter class powers.
      #   @return [ActiveRecord::Relation<Poder>]
      scope :poderes_classe, -> { where(type: "poder_classe") }

      # @!method poderes_gerais
      #   Filter general powers.
      #   @return [ActiveRecord::Relation<Poder>]
      scope :poderes_gerais, -> { where(type: "poder_geral") }

      # @!method by_type(type)
      #   Filter powers by type.
      #   @param type [String] Power type
      #   @return [ActiveRecord::Relation<Poder>]
      scope :by_type, ->(t) { where(type: t) }

      # @!method by_origin(origin_id)
      #   Filter powers by associated origin.
      #   @param origin_id [String] Origin ID
      #   @return [ActiveRecord::Relation<Poder>]
      scope :by_origin, ->(origin_id) { where(origin_id: origin_id) }

      # @!method by_deity(deity_id)
      #   Filter powers granted by a specific deity.
      #   @param deity_id [String] Deity ID
      #   @return [ActiveRecord::Relation<Poder>]
      scope :by_deity, ->(deity_id) { where("json_extract(deities, '$') LIKE ?", "%#{deity_id}%") }

      # @!endgroup

      # Convert power to a hash representation.
      #
      # @return [Hash] Hash with all power attributes
      def to_h
        {
          id: id,
          name: name,
          type: type,
          description: description,
          effects: effects,
          prerequisites: prerequisites,
          origin_id: origin_id,
          deities: deities
        }.compact
      end
    end
  end
end

File: src/ts/models/poder.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"
import { getBookReference, type BookReference } from "./concerns/book_referenceable.js"

const TABLE = "poderes"

/**
 * Valid power type identifiers.
 *
 * - `habilidade_unica_origem` — unique power granted by a character origin
 * - `poder_concedido` — power granted by a deity to a devotee
 * - `poder_tormenta` — corruption power from Tormenta exposure
 * - `poder_classe` — class feature power
 * - `habilidade_de_raca` — racial ability
 * - `poder_geral` — general power (feat)
 * - `poder_combate` — combat power
 * - `poder_destino` — destiny power
 * - `poder_magia` — magic power
 */
export type PoderType =
  | "habilidade_unica_origem" | "poder_concedido" | "poder_tormenta"
  | "poder_classe" | "habilidade_de_raca" | "poder_geral"
  | "poder_combate" | "poder_destino" | "poder_magia"

/**
 * A Poder (power or ability) in Tormenta20.
 *
 * Powers represent feats, class features, racial abilities, deity-granted
 * powers, and Tormenta corruption abilities.
 *
 * @example
 * ```ts
 * Poder.poderesCombate().all()          // all combat powers
 * Poder.byOrigin("nobre").all()         // powers from the Nobre origin
 * Poder.byDeity("khalmyr").all()        // powers granted by Khalmyr
 * Poder.find("arma_do_tempo")           // find by id
 * ```
 */
export class Poder extends BaseModel {
  get type(): PoderType { return this._row.type as PoderType }
  get description(): string | null { return (this._row.description as string) ?? null }
  /** JSON array of effect objects. Structure varies by type — see `PoderType` docs. */
  get effects(): unknown[] { return BaseModel.json(this._row.effects, []) }
  /** JSON array of activation cost descriptors (PM costs, etc.). */
  get costs(): unknown[] { return BaseModel.json(this._row.costs, []) }
  /** JSON array of prerequisite descriptors. */
  get prerequisites(): unknown[] { return BaseModel.json(this._row.prerequisites, []) }
  /** ID of the associated {@link Origem}, or `null` for non-origin powers. */
  get originId(): string | null { return (this._row.origin_id as string) ?? null }
  /** ID of the associated {@link Classe}, or `null` for non-class powers. */
  get classId(): string | null { return (this._row.class_id as string) ?? null }
  /** IDs of deities that grant this power (for `poder_concedido`). */
  get deities(): string[] { return BaseModel.json(this._row.deities, []) }

  /** Returns the sourcebook page reference, or `null` if not indexed. */
  bookReference(): BookReference | null { return getBookReference(this.id) }

  toH() {
    return {
      id: this.id, name: this.name, type: this.type, description: this.description,
      effects: this.effects, costs: this.costs, prerequisites: this.prerequisites,
      origin_id: this.originId, class_id: this.classId, deities: this.deities,
    }
  }

  static query(): Query<Poder> { return BaseModel.makeQuery(TABLE, Poder) }
  static all(): Poder[] { return Poder.query().all() }
  static find(id: string): Poder | null { return Poder.query().find(id) }
  static first(): Poder | null { return Poder.query().first() }
  static last(): Poder | null { return Poder.query().last() }
  static count(): number { return Poder.query().count() }

  /** Returns powers matching the given type. */
  static byType(type: PoderType): Query<Poder> { return Poder.query().where("type = ?", type) }
  /** Returns powers associated with the given origin ID. */
  static byOrigin(originId: string): Query<Poder> { return Poder.query().where("origin_id = ?", originId) }
  /** Returns powers associated with the given class ID. */
  static byClass(classId: string): Query<Poder> { return Poder.query().where("class_id = ?", classId) }
  /** Returns powers granted by the given deity ID. */
  static byDeity(deityId: string): Query<Poder> { return Poder.query().where("deities LIKE ?", `%${deityId}%`) }

  static habilidadesUnicas(): Query<Poder> { return Poder.byType("habilidade_unica_origem") }
  static poderesConcedidos(): Query<Poder> { return Poder.byType("poder_concedido") }
  static poderesClasse(): Query<Poder> { return Poder.byType("poder_classe") }
  static poderesGerais(): Query<Poder> { return Poder.byType("poder_geral") }
  static habilidadesDeRaca(): Query<Poder> { return Poder.byType("habilidade_de_raca") }
  static poderesTormenta(): Query<Poder> { return Poder.byType("poder_tormenta") }
  static poderesCombate(): Query<Poder> { return Poder.byType("poder_combate") }
  static poderesDestino(): Query<Poder> { return Poder.byType("poder_destino") }
  static poderesMagia(): Query<Poder> { return Poder.byType("poder_magia") }
}
